packages/assets-controllers/src/TokenRatesController.ts

```diff
@@ -421,194 +367,195 @@ export class TokenRatesController extends StaticIntervalPollingController<TokenR
-          return marketData;
-        } catch (error: unknown) {
-          updateFailed(error);
-          throw error;
-        } finally {
-          // Cleanup the tracking for this update.
-          delete this.#inProcessExchangeRateUpdates[updateKey];
-        }
-      },
-    );
+    await Promise.allSettled(promises);
 
-    // Wait for all update promises to settle.
-    const results = await Promise.allSettled(updatePromises);
+    const chainIds = new Set(
+      Object.values(chainIdAndNativeCurrency).map((chain) => chain.chainId),
+    );
 
-    // Merge all successful market data updates into one object.
-    const combinedMarketData = results.reduce((acc, result) => {
-      if (result.status === 'fulfilled' && result.value) {
-        acc = { ...acc, ...result.value };
+    for (const chainId of chainIds) {
+      if (!marketData[chainId]) {
+        marketData[chainId] = {};
       }
-      return acc;
-    }, {});
+    }
 
-    // Call this.update only once with the combined market data to reduce the number of state changes and re-renders
-    if (Object.keys(combinedMarketData).length > 0) {
+    if (Object.keys(marketData).length > 0) {
       this.update((state) => {
         state.marketData = {
           ...state.marketData,
-          ...combinedMarketData,
+          ...marketData,
         };
       });
     }
   }
 
-  /**
-   * Uses the token prices service to retrieve exchange rates for tokens in a
-   * particular currency.
-   *
-   * If the price API does not support the given chain ID, returns an empty
-   * object.
-   *
-   * If the price API does not support the given currency, retrieves exchange
-   * rates in a known currency instead, then converts those rates using the
-   * exchange rate between the known currency and desired currency.
-   *
-   * @param args - The arguments to this function.
-   * @param args.tokenAddresses - Addresses for tokens.
-   * @param args.chainId - The EIP-155 ID of the chain where the tokens live.
-   * @param args.nativeCurrency - The native currency in which to request
-   * exchange rates.
-   * @returns A map from token address to its exchange rate in the native
-   * currency, or an empty map if no exchange rates can be obtained for the
-   * chain ID.
-   */
-  async #fetchAndMapExchangeRates({
-    tokenAddresses,
-    chainId,
-    nativeCurrency,
-  }: {
-    tokenAddresses: Hex[];
-    chainId: Hex;
-    nativeCurrency: string;
-  }): Promise<ContractMarketData> {
-    if (!this.#tokenPricesService.validateChainIdSupported(chainId)) {
-      return tokenAddresses.reduce((obj, tokenAddress) => {
-        obj = {
-          ...obj,
-          [tokenAddress]: undefined,
-        };
-
-        return obj;
-      }, {});
-    }
+  async #fetchAndMapExchangeRatesForSupportedNativeCurrency(
+    assets: {
+      chainId: Hex;
+      tokenAddress: Hex;
+    }[],
+    currency: string,
+    marketData: Record<Hex, Record<Hex, MarketDataDetails>> = {},
```

Comment:
Controversial, but passing the global marketData object and mutate it directly saves us juggling with the returned results quite a bit.

Doing this is safe with parallel calls because the entry for the chain (first key) is created safely with `??=` and the token address (second key) will not be repeated in different batches.